src/shared/utilities/messages.ts

```diff
@@ -182,11 +183,114 @@ export async function showMessageWithCancel(
     message: string,
     timeout: Timeout,
     window: Window = globals.window
-): Promise<vscode.Progress<{ message?: string; increment?: number }>> {
+): Promise<vscode.Progress<ProgressEntry>> {
     const progressOptions = { location: vscode.ProgressLocation.Notification, title: message, cancellable: true }
     return showProgressWithTimeout(progressOptions, timeout, window)
 }
 
+/**
+ * The message ID that {@link CancellableMessageManager} uses.
+ * It uniquely identifies a specific message.
+ */
+export type CancellableMessageId = string
+
+/**
+ * Provides support to create cancellable vscode message windows.
+ *
+ * To get started, you create a new message with a unique ID.
+ * Following that, you can easily update the message or cancel it.
+ *
+ * ---
+ *
+ * This class solves an issue of needing to manage duplicated
+ * messages, the interface simplifies the process.
+ */
+export class CancellableMessageManager {
+    static readonly timeoutMillis = 60000
+    private static messageMap: {
+        [msgId: CancellableMessageId]: { timeout: Timeout; progress: vscode.Progress<ProgressEntry> }
+    } = {}
+
+    /**
+     * Helper to generate unique ID from the given info
+     */
+    static idFromCwlInfo(
+        cwlGroup: CloudWatchLogsGroupInfo,
+        cwlParameters: CloudWatchLogsParameters
+    ): CancellableMessageId {
+        return `${getStringHash(JSON.stringify({ cwlGroup, cwlParameters }))}`
+    }
```

Comment:
I think in terms of discoverability it will be difficult to find. What about a separate class in `messages.ts` named `CancellableMessageIdBuilder` that holds separate implementations for different modules.

```typescript
class CancellableMessageBuilder {
  static fromCwlInfo(
          cwlGroup: CloudWatchLogsGroupInfo,
          cwlParameters: CloudWatchLogsParameters
    ): CancellableMessageId {
        return `${getStringHash(JSON.stringify({ cwlGroup, cwlParameters }))}`
    }
}

Reply:
But why does the id have to be specific to cancellable messages? The identifier is provided by the _caller_ and is completely unconstrained. Let's stay focused on the purpose of the string and not how it's created. In this case, isn't our goal to limit cancel notifications to a single URI? So why not use `uriToKey` in `src/cloudWatchLogs/cloudWatchLogsUtils.ts`? Needing a unique key for something is not an uncommon problem, and we try to re-use existing formats when it makes sense to do so.

Reply:
Thanks for pointing out the utils. I've updated to use `uriToKey()` when we have the uri and `uriToKey(createURIFromArgs(logGroupInfo, parameters))` for another section.